---
title: 'ESM text for ldeli growth'
author: Fonti Kar
date: "`r Sys.Date()`"
output: 
  word_document:
    reference_docx: "Style_guide.docx"
---

```{r setup, include=FALSE, eval = T, echo = FALSE, warning = FALSE}
knitr::opts_chunk$set(echo = TRUE)
rm(list=ls())
sessionInfo()

# R version 3.6.1 (2019-07-05)
# Platform: x86_64-apple-darwin15.6.0 (64-bit)
# Running under: macOS High Sierra 10.13.6

pacman::p_load(dplyr, magrittr, ggplot2, patchwork, lubridate, stringr, tidyr, MCMCglmm, bookdown, pander, brms)

#Read in dataset
#Complete
data <- read.csv("~/Dropbox/1 - PhD/4 - ldeli_growthrate/data/growth/processed/analysis/Ldeli_quangen_growth.csv", stringsAsFactors = F)

#Augmented
data_DA <- read.csv("~/Dropbox/1 - PhD/4 - ldeli_growthrate/data/growth/processed/analysis/Ldeli_quangen_growth_DA.csv", stringsAsFactors = F)

```

## Pedigree and genomic relatedness

We submitted a total of 437 tissues samples, five samples experienced problems during extraction and sequencing and were therefore excluded from the final dataset (n = 432). 

DNA was extracted from tissue samples using a Qiagen DNeasy Blood and Tissue Kits following the manufacturer's instructions. Diversity Arrays Technology (DArT) combines next generation sequencing platforms and genome complexity reduction methods (Kilian et al., 2012) to select the most appropriate method for L.delicata.
[Insert section from DArT here on complexity reduction methods, DArT tested four methods in Gambusia so it's unclear what method was used for us].

Sequences from all lanes were then processed using DArT specific pipelines. The main pipeline filtered out poor quality sequences [Details needed from DArT]. For our samples, this filtering process by DArT resulted in a total of 185,963 SNPs. 

One individual was excluded from the dataset and may be related to contamination as this individual appeared to be unrelated to any other samples in the dataset.

## Model fitting and selection of random effects structure
```{r, echo = FALSE, warning = FALSE}
#Read in models
mod.4.1.DA <- readRDS("~/Dropbox/1 - PhD/4 - ldeli_growthrate/output/rds/mod.4.1.DA")
mod.4.2.DA <- readRDS("~/Dropbox/1 - PhD/4 - ldeli_growthrate/output/rds/mod.4.2.DA")
mod.4.3.DA <- readRDS("~/Dropbox/1 - PhD/4 - ldeli_growthrate/output/rds/mod.4.3.DA")
mod.4.4.DA <- readRDS("~/Dropbox/1 - PhD/4 - ldeli_growthrate/output/rds/mod.4.4.DA")
mod.4.6.DA <- readRDS("~/Dropbox/1 - PhD/4 - ldeli_growthrate/output/rds/mod.4.6.DA")
mod.4.5.DA <- readRDS("~/Dropbox/1 - PhD/4 - ldeli_growthrate/output/rds/mod.4.5.DA")

re_dic <- data.frame(matrix(nrow = 6, ncol = 2))
colnames(re_dic) <- c("formula", "DIC")

m.4.5 <- summary(mod.4.5.DA)$random.formula #null
m4.1 <- summary(mod.4.1.DA)$random.formula #quadractic all
m4.2 <- summary(mod.4.2.DA)$random.formula #linear all 

m4.4 <- summary(mod.4.4.DA)$random.formula #fix liz_id (G)
m4.3 <- summary(mod.4.3.DA)$random.formula #fix dam
m4.6 <- summary(mod.4.6.DA)$random.formula #fix PE


re_dic[1,1] <- c("intercepts for liz_id, dam_id, id") #format(m.4.5)
re_dic[2,1] <-  c("quadratic slopes for liz_id, dam_id, id") #paste0(format(m4.1)[1], trimws(format(m4.1)[2]), trimws(format(m4.1)[3]))
re_dic[3,1] <-  c("linear slopes for liz_id, dam_id, id") #paste0(format(m4.2)[1], trimws(format(m4.2)[2]))
re_dic[4,1] <- c("intercept for liz_id, linear slopes for dam_id, id") #format(m4.4)
re_dic[5,1] <-  c("intercept for dam_id, linear slopes for liz_id, id") #format(m4.3)[1]
re_dic[6,1] <- c("intercept for id, linear slopes for liz_id, dam_id")#paste0(format(m4.6)[1], trimws(format(m4.6)[2]))

re_dic[1,2] <- mod.4.5.DA$DIC
re_dic[2,2] <- mod.4.1.DA$DIC
re_dic[3,2] <- mod.4.2.DA$DIC
re_dic[4,2] <- mod.4.4.DA$DIC
re_dic[5,2] <- mod.4.3.DA$DIC
re_dic[6,2] <- mod.4.6.DA$DIC
```

```{r, echo = FALSE}
pander(re_dic,
       justify = "lc",
       split.cell = 80,
       caption = "S1 DIC of six intercept models with different combinations of random effects structures for additive genetic variance, maternal variance and permanent environmental variance")
```

## Do growth tracjectories differ among incubation treatments?

The first model we fitted included an interaction between treatment and the linear and quadratic component of the mass growth curve in the fixed effects. We then subsequently dropped non-signficant interactions in a stepwise manner. The second model only had an interaction between treatment and the linear component of the growth curve and finally the third model had the main effect of treatment in the model and no interactions with the growth curve components. Coefficient estimates from all three models are provided in Table. S2-4.
```{r, echo = FALSE, warning = FALSE}
#Read in models
mod.5.DA <- readRDS("~/Dropbox/1 - PhD/4 - ldeli_growthrate/output/rds/mod.5.DA")

#Combining the coefficients from the final model
mod.5.DA.fixed <- posterior_summary(mod.5.DA$Sol)[,-2] %>% round(3) 

mod.5.DA.liz_id <- posterior_summary(mod.5.DA$VCV)[c(1,5,9,2,3,6),-2] %>% round(3)   #liz_id random effects

mod.5.DA.dam_id <- posterior_summary(mod.5.DA$VCV)[c(10,14,18,11,12,15),-2] %>% round(3) #dam_id random effects

mod.5.DA.pe <- posterior_summary(mod.5.DA$VCV)[19,-2] %>% round(3) #pe random effects

mod.5.DA.units <- posterior_summary(mod.5.DA$VCV)[20,-2] %>% round(3) #units

#Create a table
mod.5.DA.coeffs <- as.data.frame(matrix(ncol = 4, nrow = 20))
colnames(mod.5.DA.coeffs) <- c("Variable", "Estimate", "Lower", "Upper")

#Create Variable names
mod.5.DA.coeffs$Variable[1:6] <- rownames(mod.5.DA.fixed)
mod.5.DA.coeffs$Variable[7:12] <- rownames(mod.5.DA.liz_id)
mod.5.DA.coeffs$Variable[13:18] <- rownames(mod.5.DA.dam_id)
mod.5.DA.coeffs$Variable[19] <- "V_pe"
mod.5.DA.coeffs$Variable[20] <- "Residuals"

#Fill in table
mod.5.DA.coeffs[1:6, 2:4] <- mod.5.DA.fixed
mod.5.DA.coeffs[7:12, 2:4] <- mod.5.DA.liz_id
mod.5.DA.coeffs[13:18, 2:4] <- mod.5.DA.dam_id
mod.5.DA.coeffs[19, 2:4] <- mod.5.DA.pe
mod.5.DA.coeffs[20, 2:4] <- mod.5.DA.units

#Duplicate table to make nicer variable names for table
Table_S2 <- mod.5.DA.coeffs

#Create Variable names
Table_S2$Variable[1:6] <- c("Intercept", "Treatment", "Age", "Age^2", "Treatment*Age", 
                           "TreatmentAge^2")
Table_S2$Variable[7:12] <- c("V_g intercept", "V_g  slope", "V_g  curvature", "Cov_g intercept - slope", 
                            "Cov_g intercept - curvature", "Cov_g slope - curvature")

Table_S2$Variable[13:18] <- c("V_m intercept", "V_m  slope", "V_m curvature", "Cov_m intercept - slope", 
                            "Cov_m intercept - curvature", "Cov_m slope - curvature")
```

```{r, echo = FALSE}
pander(Table_S2,
       justify = "lccc",
       #split.cell = 80,
       caption = "S2 Model coefficients from the full model where incubation temperature interacts with age and age^2. Response was log transformed. All fixed effects were z-transformed")
```

```{r, echo = FALSE, warning = FALSE}
mod.5.1.DA <- readRDS("~/Dropbox/1 - PhD/4 - ldeli_growthrate/output/rds/mod.5.1.DA")

#Combining the coefficients from the final model
mod.5.1.DA.fixed <- posterior_summary(mod.5.1.DA$Sol)[,-2] %>% round(3) 

mod.5.1.DA.liz_id <- posterior_summary(mod.5.1.DA$VCV)[c(1,5,9,2,3,6),-2] %>% round(3)   #liz_id random effects

mod.5.1.DA.dam_id <- posterior_summary(mod.5.1.DA$VCV)[c(10,14,18,11,12,15),-2] %>% round(3) #dam_id random effects

mod.5.1.DA.pe <- posterior_summary(mod.5.1.DA$VCV)[19,-2] %>% round(3) #pe random effects

mod.5.1.DA.units <- posterior_summary(mod.5.1.DA$VCV)[20,-2] %>% round(3) #units

#Create a table
mod.5.1.DA.coeffs <- as.data.frame(matrix(ncol = 4, nrow = 19))
colnames(mod.5.1.DA.coeffs) <- c("Variable", "Estimate", "Lower", "Upper")

#Create Variable names
mod.5.1.DA.coeffs$Variable[1:5] <- rownames(mod.5.1.DA.fixed)
mod.5.1.DA.coeffs$Variable[6:11] <- rownames(mod.5.1.DA.liz_id)
mod.5.1.DA.coeffs$Variable[12:17] <- rownames(mod.5.1.DA.dam_id)
mod.5.1.DA.coeffs$Variable[18] <- "V_pe"
mod.5.1.DA.coeffs$Variable[19] <- "Residuals"

#Fill in table
mod.5.1.DA.coeffs[1:5, 2:4] <- mod.5.1.DA.fixed
mod.5.1.DA.coeffs[6:11, 2:4] <- mod.5.1.DA.liz_id
mod.5.1.DA.coeffs[12:17, 2:4] <- mod.5.1.DA.dam_id
mod.5.1.DA.coeffs[18, 2:4] <- mod.5.1.DA.pe
mod.5.1.DA.coeffs[19, 2:4] <- mod.5.1.DA.units

#Duplicate table to make nicer variable names for table
Table_S3 <- mod.5.1.DA.coeffs

#Create Variable names
Table_S3$Variable[1:5] <- c("Intercept", "Treatment", "Age", "Age^2", "Treatment*Age")

Table_S3$Variable[7:12] <- c("V_g intercept", "V_g  slope", "V_g  curvature", "Cov_g intercept - slope", 
                            "Cov_g intercept - curvature", "Cov_g slope - curvature")

Table_S3$Variable[13:18] <- c("V_m intercept", "V_m  slope", "V_m curvature", "Cov_m intercept - slope", 
                            "Cov_m intercept - curvature", "Cov_m slope - curvature")
```

```{r, echo = FALSE}
pander(Table_S3,
       justify = "lccc",
       #split.cell = 80,
       caption = "S3 Model coefficients from a model where incubation temperature interacts with age only. Response was log transformed. All fixed effects were z-transformed")
```

```{r, include = FALSE, echo = FALSE, warning = FALSE}
mod.5.2.DA <- readRDS("~/Dropbox/1 - PhD/4 - ldeli_growthrate/output/rds/mod.5.2.DA")

#Combining the coefficients from the final model
mod.5.2.DA.fixed <- posterior_summary(mod.5.2.DA$Sol)[,-2] %>% round(3) 

mod.5.2.DA.liz_id <- posterior_summary(mod.5.2.DA$VCV)[c(1,5,9,2,3,6),-2] %>% round(3)   #liz_id random effects

mod.5.2.DA.dam_id <- posterior_summary(mod.5.2.DA$VCV)[c(10,14,18,11,12,15),-2] %>% round(3) #dam_id random effects

mod.5.2.DA.pe <- posterior_summary(mod.5.2.DA$VCV)[19,-2] %>% round(3) #pe random effects

mod.5.2.DA.units <- posterior_summary(mod.5.2.DA$VCV)[20,-2] %>% round(3) #units

#Create a table
mod.5.2.DA.coeffs <- as.data.frame(matrix(ncol = 4, nrow = 18))
colnames(mod.5.2.DA.coeffs) <- c("Variable", "Estimate", "Lower", "Upper")

#Create Variable names
mod.5.2.DA.coeffs$Variable[1:4] <- rownames(mod.5.2.DA.fixed)
mod.5.2.DA.coeffs$Variable[5:10] <- rownames(mod.5.2.DA.liz_id)
mod.5.2.DA.coeffs$Variable[11:16] <- rownames(mod.5.2.DA.dam_id)
mod.5.2.DA.coeffs$Variable[17] <- "V_pe"
mod.5.2.DA.coeffs$Variable[18] <- "Residuals"

#Fill in table
mod.5.2.DA.coeffs[1:4, 2:4] <- mod.5.2.DA.fixed
mod.5.2.DA.coeffs[5:10, 2:4] <- mod.5.2.DA.liz_id
mod.5.2.DA.coeffs[11:16, 2:4] <- mod.5.2.DA.dam_id
mod.5.2.DA.coeffs[17, 2:4] <- mod.5.2.DA.pe
mod.5.2.DA.coeffs[18, 2:4] <- mod.5.2.DA.units

#Duplicate table to make nicer variable names for table
Table_S4 <- mod.5.2.DA.coeffs

#Create Variable names
Table_S4$Variable[1:4] <- c("Intercept", "Treatment", "Age", "Age^2")

Table_S4$Variable[5:10] <- c("V_g intercept", "V_g  slope", "V_g  curvature", "Cov_g intercept - slope", 
                            "Cov_g intercept - curvature", "Cov_g slope - curvature")

Table_S4$Variable[11:16] <- c("V_m intercept", "V_m  slope", "V_m curvature", "Cov_m intercept - slope", 
                            "Cov_m intercept - curvature", "Cov_m slope - curvature")

pander(Table_S4,
       justify = "lccc",
       #split.cell = 80,
       caption = "S3 Model coefficients from the main effects model. Response was log transformed. All fixed effects were z-transformed")
```

## Missing data augmentation using MCMCglmm

```{r, echo = FALSE, include = FALSE}
#Number of row for each individual
row_data <- data %>% group_by(liz_id) %>% 
  summarise(nrows = length(measure_date),) %>% 
  arrange(-nrows) %>% as.data.frame()

#Number of rows needed to be added for some individuals 
row_data %>% filter(!nrows >= 16) %>% mutate(rows_add = 16 - nrows) #working with 208 individuals
row_data %<>% filter(!nrows >= 16) %>% mutate(rows_add = 16 - nrows)

#Despritive stats on missing data
mean(row_data$rows_add)%>% round(2)

nrow(data_DA) - nrow(data)
```

We wanted to maximise the number of individuals in the dataset and not exclude lizards died before we attained enough data for their full growth curves (n = `r row_data %>% filter(!nrows >= 16) %>% nrow()`). We made use of the built-it capabilities of missing data augmention in the R package 'MCMCglmm'. We achieved this by creating empty rows of data for lizards that we had < 16 measurements. First we determined how many growth measurements a given lizard is missing (mean missing = `r mean(row_data$rows_add)%>% round(2)`). Based on this information, we then generated additionals rows of data for each lizard. For each additional row of data, we set a 35 day interval for 'days since hatch/age' relative to the last known growth measurement of a given lizard. Mass, SVL, lnMass and lnSVL. are set as NA in these additional rows. We then created an additional variable which represents the curvature component of the growth curve by squaring 'days since hatch/age'. A total of `r nrow(data_DA) - nrow(data)` rows were added to this dataset.We fitted the final models with the augmented dataset. See provided code for more details.

Conclusions of our results did not differ between using the original dataset vs. the augmented dataset (Table X -X, Fig X - X). Paramater estimates and credible intervals overlapped, we therefore present the results from models using the augmented dataset

```{r}

```








